### DISCUSSION TITLE
Is it possible to build a `regex::Regex` from a `regex_automata::meta::Regex`?

### DISCUSSION THREAD

**Discussion by xmo-odoo** [rust-lang/regex#1197] [Q&A] [answered]
I'm using `regex_syntax` to work with HIR then converting the HIR to a `regex_automata` for matching proper, however to expose to users `regex_automata` might be a bit low-level and involved (especially when it comes to captures), so I wanted to know if there was an easy way to convert into a `Regex` with all the convenient high-level helpers, or if I'd have to either expose the lower level APIs or create my own convenience wrapper.

**Comment by BurntSushi** [comment, marked as answer]
Good question. This is similar to questions folks have asked in the past (before `regex-automata` existed) like, "If I have an `Hir` from `regex-syntax`, can I use that to directly build a `regex::Regex`?" The answer to that is basically the same answer to your question: definitively and intentionally not. The only way for a conversion like the one you're asking for to exist is if `regex-automata` were a public dependency of `regex`. That in turn would mean that every time `regex-automata` had a semver incompatible release, `regex` would also need a semver compatible release. This is not something I can abide, because it would imply that `regex-automata` could not evolve independently of `regex`. That is, it would inhibit its primary goal. From the docs (emphasis mine):

> **The primary goal of this crate is to serve as an implementation detail for the regex crate.** A secondary goal is to make its internals available for use by others.

With that said, not all hope is lose here because there _is_ a stable interface between `regex-syntax` and `regex`: the concrete syntax. If you're working with `Hir`, then you can always call `hir.to_string()` and pass it to `Regex::new`. ripgrep used this for years before it switched over to `regex-automata` for its matching. (ripgrep doesn't even use `regex` directly.) In particular, the conversion from `Hir` to a string, and then to a regex via `Regex::new` is guaranteed to work.

It does sound wasteful, and technically, it is. But necessarily so given the constraints above. With that said, it is _probably_ not as wasteful as you might think. In particular, the parsing and creation of an `Hir` would be repeated, but this is one of the least expensive parts of `Regex::new`. It is _much_ more expensive to build the internal finite state machine data structures, and that is not being repeated by the following process: `parse -> HIR -> modify HIR -> to string -> Regex::new`.

**Reply by BurntSushi** [reply]
Also, the only reason why you can build a `meta::Regex` from an `Hir` is because `regex-syntax` is a public dependency of `regex-automata`. This is okay because `regex-syntax` has been baking for a long time and I generally expect it to evolve more slowly than `regex-automata`.

**Reply by xmo-odoo** [reply]
Thanks, that seems fair. I don't actually modify the HIR (only parse / datamine it), so I can convert all the regex-automata bit to straight `regex`, and re-parse the source expression, with some sort of bridge for the flags I might be using so they're used for both HIR parsing and regex.
